config/tools/config_generator.cpp

```diff
@@ -0,0 +1,897 @@
+        }
+        double percentage_overlapped_so_far
+            = (total_sum
+               / static_cast<double>(shard_info.at(shard_id).numbers_covered))
+            - 1;
+        if(shard_info.at(shard_id).overlap_percentage_allowed
+               <= percentage_overlapped_so_far
+           || ((shard_info.at(shard_id)
+                    .current_coverage_expansion_limits.second
+                - shard_info.at(shard_id)
+                      .current_coverage_expansion_limits.first)
+               == (array_total.size() - 1))) {
+            // This shard has reached its allowed overlap limit
+            shard_info.at(shard_id).still_expanding = false;
+        }
+        // Check if all indices have been visited at least once from 0 to
+        // shard_size
+        bool all_visited = true;
+        for(size_t i = 0; i < array_total.size(); i++) {
+            if(array_total.at(i) == 0) {
+                all_visited = false;
+                break;
+            }
+        }
+        // If all indices have been visited at least once from 0 to
+        // shard_size, allow shards to start overlapping in coverage
+        if(all_visited == true) {
+            for(std::vector<ShardInfo>::iterator shard_it = shard_info.begin();
+                shard_it != shard_info.end();
+                shard_it++) {
+                shard_it->allow_overlap = true;
+            }
+        }
+    }
+
+    // Get random value based on mean and standard deviation
+    [[nodiscard]] auto
+    config_generator::calculate_normal_distribution_point(const size_t mean,
+                                                          const double std_dev,
+                                                          const bool randomize)
+        -> double {
+        if(randomize) {
+            generator.seed(
+                std::chrono::system_clock::now().time_since_epoch().count());
+        }
+        std::normal_distribution<double> distribution(mean, std_dev);
+        return distribution(generator);
+    }
+
+    // This is not a failsafe because we are simply generating a
+    // configuration file here, however, it is still good to check that the
+    // port is available
+    [[nodiscard]] auto config_generator::get_open_port() -> unsigned short {
+        unsigned short port = m_current_port % MAX_PORT_NUM;
+        m_current_port++;
+        auto ep = cbdc::network::endpoint_t{cbdc::network::localhost, port};
+        auto listener = cbdc::network::tcp_listener();
+        while(!listener.listen(ep.first, ep.second)) {
+            port = m_current_port % MAX_PORT_NUM;
+            ep = cbdc::network::endpoint_t{cbdc::network::localhost, port};
+            m_current_port++;
+        }
+        return port;
+    }
+
+    [[nodiscard]] auto config_generator::create_repeatable_key_pair()
+        -> std::pair<std::string, std::string> {
+        cbdc::privkey_t seckey;
+        for(auto&& b : seckey) {
+            b = rand();
```

Comment:
It is repeatable depending on the template configuration. If you dont want it to be repeatable, if you look around line 92 (in the constructor), you will see that it will use srand to set the seed with a value related to the current time. If not set to randomize, srand is not called and the default value is used and is thus repeatable from run to run.

NEEDS FOLLOW UP

Reply:
Ahh, I see. It's depending on `rand()` behaving as if `srand(1)` had been called. That's reasonable (and specified by the standard), but escaped me in-review. I guess my next question is why call `srand()` ever then? There's only one other place this code uses `rand()` and it doesn't look like it has any real need to be particularly random (if it does, replacing the call to `rand()` with the more modern tooling is the better option in my opinion).

Reply:
Got it. That makes sense yes. But on the other hand, this is just for creating configuration files for tests. I can change it out though no problem if you want.